rdmalib/include/rdmalib/connection.hpp

```diff
@@ -160,27 +194,106 @@ namespace rdmalib {
     }
-    int32_t post_atomic_fadd(const Buffer<uint64_t> & _accounting_buf, const RemoteBuffer & rbuf, uint64_t add);
-    #else
-    int32_t post_atomic_fadd(ScatterGatherElement && elems, const RemoteBuffer & rbuf, uint64_t add);
-    #endif
+
+    int32_t post_atomic_fadd(const Buffer<uint64_t> & _accounting_buf, const RemoteBuffer_t & rbuf, uint64_t add);
 
     // Register to be notified about all events, including unsolicited ones
-    #ifdef USE_LIBFABRIC
     int wait_events(int timeout = -1);
-    #else
+
+    int32_t _post_write(ScatterGatherElement_t && elems, const RemoteBuffer_t & rbuf, const uint32_t immediate = 0);
+    int32_t post_write(ScatterGatherElement_t && elems, const RemoteBuffer_t & rbuf, bool force_inline);
+
+    std::tuple<fi_cq_data_entry *, int> poll_wc(QueueType type, bool blocking=true, int count=-1, bool update=false);
+  };
+
+  struct VerbsConnection : Connection<VerbsConnection, ibverbs>
+  {
+    using Library = ibverbs;
+    id_t _id;
+    channel_t _channel;
+
+    struct ibv_recv_wr _batch_wrs[_rbatch]; // preallocated and prefilled batched recv.
+    std::array<ScatterGatherElement_t, _wc_size> _rwc_sges;
+
+    VerbsConnection(bool passive=false);
+    VerbsConnection(VerbsConnection&& obj);
+    ~VerbsConnection();
+    void close();
+
+    id_t id() const
+    {
+      return this->_id;
+    }
+
+    void inlining(bool enable);
+    template <typename B>
+    void initialize_batched_recv(const rdmalib::impl::Buffer<B, ibverbs> & sge, size_t offset);
+    void initialize(rdma_cm_id* id);
+    ibv_comp_channel* completion_channel() const;
+
+    int32_t post_send(const ScatterGatherElement_t & elem, int32_t id = -1, bool force_inline = false);
+    int32_t post_batched_empty_recv(int count);
+    int32_t post_recv(ScatterGatherElement_t && elem, int32_t id, int count=1);
+    int32_t post_cas(ScatterGatherElement_t && elems, const RemoteBuffer_t & rbuf, uint64_t compare, uint64_t swap);
+    int32_t post_atomic_fadd(ScatterGatherElement_t && elems, const RemoteBuffer_t & rbuf, uint64_t add);
+    int32_t post_atomic_fadd(const Buffer<uint64_t, Library> & _accounting_buf, const RemoteBuffer_t & rbuf, uint64_t add);
+
     void notify_events(bool only_solicited = false);
     ibv_cq* wait_events();
     void ack_events(ibv_cq* cq, int len);
-    #endif
-  private:
-    #ifdef USE_LIBFABRIC
-    int32_t _post_write(ScatterGatherElement && elems, const RemoteBuffer & rbuf, const uint32_t immediate = 0);
-    #else
-    int32_t _post_write(ScatterGatherElement && elems, ibv_send_wr wr, bool force_inline, bool force_solicited);
-    #endif
+
+    int32_t _post_write(ScatterGatherElement_t && elems, ibv_send_wr wr, bool force_inline, bool force_solicited);
+    int32_t post_write(ScatterGatherElement_t && elems, const RemoteBuffer_t & rbuf, bool force_inline);
+
+    std::tuple<ibv_wc*, int> poll_wc(QueueType type, bool blocking=true, int count=-1);
+
   };
+
+  template <typename B>
```

Comment:
A standard convention for a generic type is letter `T` btw.

Reply:
Okay, noted, thanks.